src/bap_mc/bap_mc.ml

```diff
@@ -103,30 +110,41 @@ let print_bil lift bil_formats mem insn =
   List.iter bil_formats ~f:(fun fmt ->
       printf "%s@." (string_of_bil fmt (bil insn)))
 
-let make_print lift insn_fmt bil_fmt show_kinds mem insn =
-  print_insn insn_fmt insn;
+let make_print lift insn_fmt bil_fmt show_kinds oplen_bases mem insn =
+  print_insn_oplen oplen_bases mem;
+  print_insn insn_fmt insn; 
   print_bil lift bil_fmt mem insn;
   if show_kinds then print_kinds insn
 
-let step print state mem insn _ =
-  print mem insn;
+
+let check disasm_len current_total =
+  if disasm_len=0 then true
+  else current_total < disasm_len
+
+let step disasm_len hit_count print state mem insn _ =
+  if (check disasm_len !hit_count) then (
+    hit_count := !hit_count+1;
+    print mem insn;
+  ) else ();
   Dis.step state (Dis.addr state |> Addr.to_int |> ok_exn)
 
-let disasm src arch show_insn show_bil show_kinds =
+let disasm src offset disasm_len arch show_oplen show_insn show_bil show_kinds =
   let arch = match Arch.of_string arch with
     | None -> raise Unknown_arch
     | Some arch -> arch in
   let module Target = (val target_of_arch arch) in
-  let print = make_print Target.lift show_insn show_bil show_kinds in
+  let print = make_print Target.lift show_insn show_bil show_kinds show_oplen in
   let input = read_input src in
   Dis.create ~backend:"llvm" (Arch.to_string arch) >>= fun dis ->
-  let invalid state mem off = no_disassembly state off in
+  let invalid state mem off = no_disassembly state off offset in
+  let hit_count = ref 0 in
```

Comment:
There is no need to use a reference here. The last parameter of this function is specially reserved for user data. Currently it is the address, where disassembler machine is pointing, that is used to check, that we consumed all data. You can extend the state to tuple, of address and counter, and increment it in each step.